### ISSUE TITLE
LuisRecognizer lacks complete typing

### ISSUE THREAD

**Issue by ryanvolum** [microsoft/botbuilder-js#682]
## Versions
SDK Version: 4.1.5
Node Version: v10.13.0
OS: Windows

## Describe the bug
When following the LUIS sample, it indicates that to get the top scoring intent you can do the following: 
```
const results = await this.luisRecognizer.recognize(turnContext);
const topIntent = results.luisResult.topScoringIntent;
```
However, when inspecting the code, neither `luisResult` nor `topScoringIntent` seem to be part of the LUIS result. Intellisense therefore doesn't show it an option. 

![image](https://user-images.githubusercontent.com/3891951/49481959-1e6e1300-f7e2-11e8-8efd-6ea74944f2b5.png)
Additionally, the `entities` property on the luis response is typed as `any`. This is especially problematic for TS developers. 

## To Reproduce
1. Follow LUIS sample
2. Examine the result of a call against luisRecognizer.recognize

[bug]

**Comment by Zerryth** [comment]
:eyes:

**Comment by Zerryth** [comment]
@ryanvolum I'm not finding the lack of correct Intellisense as you've described.

I do notice that you're using an older version `4.1.5`, however when you clone the samples today, they actually point to `botbuilder-ai 4.2.0`.

![image](https://user-images.githubusercontent.com/35248895/52001302-edbdf900-2473-11e9-920a-614403423056.png)

Try recloning the samples and so in turn when you `npm i` everything it should install up-to-date packages.

**Comment by Zerryth** [comment]
Closing, as this issue was raised back in early Dec, and it appears to be fixed as of late Jan

**Comment by ctstone** [comment]
This is still an issue, at least for entities. [RecognizerResult](https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botbuilder-core/src/recognizerResult.ts#L34) defines the `entities` property as type `any`. This is definitely confusing because `LuisRecognizer` defines its own entity structure separate from the direct LUIS API.

**Comment by ryanvolum** [comment]
@johnataylor, @Zerryth I'm not sure how the text and the images provided lack correlation. There's no `topScoringIntent` off of the `RecognizerResult` and entities _are_ typed as any:

![image](https://user-images.githubusercontent.com/3891951/55973348-37784c00-5c3a-11e9-8511-4176a64fe00a.png)

To @ctstone's point, this confuses people because they have no idea what a `LuisRecognizer` entity is or how to use it.

Note that I'm using latest: `"botbuilder-ai": "^4.3.4"`

**Comment by Zerryth** [comment]
hey gang I have it on my to-do list to investigate this again (need to wrap up some other tasks first)

I'll update thread to new findings/questions then though

**Comment by Zerryth** [comment]
#### Accessing `topIntent`
So this issue was opened originally last year, and the samples and design of the `luisRecognizer` has changed since then.
* Originally, according to the OP the samples showed that you would access `topIntent` directly off the `RecognizerResult`  ![image](https://user-images.githubusercontent.com/35248895/58054109-6d003700-7b0e-11e9-9afd-579a47729a04.png)

* Now, however, if you look at [sample 14.nlp-with-dispatch](https://github.com/microsoft/BotBuilder-Samples/blob/master/samples/javascript_nodejs/14.nlp-with-dispatch/bots/dispatchBot.js#L44), the method demonstrated to access `topIntent` would be to use the public static method off of the `LuisRecognizer` class, passing in your `RecgonizerResult` that you got from `.recognize()` as an arg:
    * `const recognizerResult = await dispatchRecognizer.recognize(context);`
    * `const intent = LuisRecognizer.topIntent(recognizerResult);`

___

#### Typing on Entities
So, as ctstone pointed out, yes the entities off of the `RecognizerResult` class are of type `any`. And yes, the SDK does structure the results from LUIS differently than just directly from the LUIS API. 

This is not unique to just `botbuilder-js`, but is also of [`JObject` type](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder/RecognizerResult.cs#L50) on the `dotnet` repo as well

As per Steve Ickman's suggestion, pinging @chrimc62 to comment on the design, and whether or not there can be anything done to minimize the confusion between the discrepancies:

> This is still an issue, at least for entities. RecognizerResult defines the entities property as type any. This is definitely confusing because LuisRecognizer defines its own entity structure separate from the direct LUIS API.

**Comment by Zerryth** [comment]
Ok, just spoke to Chris McConnell in person, and he illuminated the down-low on the LUIS result mapping situation.

In a nutshell, it sounds like the solution to your woes would be in use of the [LUISGen tool](https://github.com/microsoft/botbuilder-tools/blob/master/packages/LUISGen/src/npm/readme.md), which is used to generate strongly typed C# classes or TS interfaces, which would make consuming LUIS output easier. This will then enable Intellisense against entities defined in your LUIS application :)
___
To add a layer of complexity to this all, yes the mapping between the BF SDK vs. the results from the LUIS API differed in structure--especially between LUIS **v2** API and the SDK.

There is an on-going effort to massage the results of the LUIS API and in the SDK to look identical to one another in **v3** of the LUIS API, which is currently in preview. Currently they are about 95% identical now, which is much improved from v2.
___
TL;DR use LUISGen to create strongly typed classes & the BF v4 SDK vs LUIS API results should more closely align in the next version (v3) of the LUIS API
